### ISSUE TITLE
prefer-destructuring proposes object destructuring for array

### ISSUE THREAD

**Issue by vsemozhetbyt** [eslint/eslint#7884]
* **ESLint Version:** 3.13.0
* **Node Version:** 7.4.0
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
'prefer-destructuring': [2, { array: true, object: true }, { enforceForRenamedProperties: true }]
```

**What did you do? Please include the actual source code causing the issue.**
```js
/******************************************************************************/
'use strict';
/******************************************************************************/
const arr = ['a', 'b', 'c'];
const last = arr[arr.length - 1];
console.log(last);
```

**What did you expect to happen?**
No error report.

**What actually happened? Please include the actual, raw output from ESLint.**
```
5:7  error  Use object destructuring  prefer-destructuring
```
The same with this code (that seems to contradict 'Examples of correct code for this rule' in the [doc](http://eslint.org/docs/rules/prefer-destructuring)).
```js
const arr = ['a', 'b', 'c'];
const lasIndex = arr.length - 1;
const last = arr[lasIndex];
console.log(last);
```
```
6:7  error  Use object destructuring  prefer-destructuring
```

**Comment by fasttime** [comment]
I believe in the first snippet, the line

```js
const last = arr[arr.length - 1];
```

can be written as

```js
const { [arr.length - 1]: last } = arr;
```

And in the second snippet

```js
const last = arr[lastIndex];
```

is the same as

```js
const { [lastIndex]: last } = arr;
```

**Comment by vsemozhetbyt** [comment]
I agree that formally Array is Object and index access can be rewritten as property access. But it seems to be a confusing and complicating suggestion.

**Comment by kaicataldo** [comment]
I honestly agree that enforcing destructuring with the computed property is confusing and not very easy to read. I would support an option to relax this rule for computed properties. Maybe `enforceForComputedProperties`, to be in line with the current schema?

```json
{
  "rules": {
    "prefer-destructuring": ["error", {
      "array": true,
      "object": true
    }, {
      "enforceForRenamedProperties": false
      "enforceForComputedProperties": false
    }]
  }
}
```

**Comment by vitorbal** [comment]
Sounds good to me. Should it be `false` by default?

**Comment by not-an-aardvark** [comment]
I think it would be reasonable to have an option to disable the rule with computed properties, but it's worth noting that this would be redundant with the `enforceForRenamedProperties` option. `enforceForRenamedProperties: false` would also imply `enforceForComputedProperties: false`, unless we change the behavior of `enforceForRenamedProperties` to not apply to computed properties.

**Comment by popkutty** [comment]
How to fix the same error for object reassignment. 
```
let { a } = conf.a;
if(conf.b) {
  a = conf.b
}
```
